### ISSUE TITLE
Don't work. NOENT ERROR

### ISSUE THREAD

**Issue by caiofacchinato** [janis91/ocr#122]
# Bug report / Feature request
## Steps to Reproduce (for bugs)

1. On my Ubuntu 17.10 (on path /var/www/html/apps/ocr/ [redis and worker, respectively])
2. docker build --build-arg PASSWORD="OCR" -t root/redis .
3. docker build --build-arg LANGUAGES="tesseract-ocr-por" -t root/ocr .
4. docker run --name redis --network=isolated_ocr -p 6379:6379 -d root/redis
5. docker run --name ocr --network=isolated_ocr -e "NODE_ENV=production" -e "REDIS_PASSWORD=OCR" -e "REDIS_HOST=redis" -e "REDIS_PORT=6379" -e "REDIS_DB=0" -v /mnt/volume-nyc1-01:/home/node/data:ro -v /var/www/html/tmp:/home/node/output -d root/ocr
6. config.php temp directory:   'tempdirectory' => '/var/www/html/tmp',
7. After enable app and try OCR a jpg, i get this msg in docker logs of ocr container: 2017-09-12T19:42:02.101Z - error: Job errored: ENOENT: no such file or directory, unlink '/home/node/output//var/www/tmp/ocr_MDEOq4.txt'
8. Error in Logging (nextcloud interface): ENOENT: no such file or directory, unlink '/home/node/output//var/www/tmp/ocr_MDEOq4.txt'
## Your Environment

* OCR version used: Your Docker version
* Browser Name and version: Chrome 60 and Opera
* Operating System and version (desktop or mobile): Server: Ubuntu 17.10 64Bits
* ownCloud/nextcloud version: (see ownCloud admin page or version.php) 12.0.0
* PHP version: 7.0.22
* Database version Lastest MariaDB
* Are you using encryption: no

## Log File Content (nextcloud/owncloud.log of the "data"-directory)
{"reqId":"SpT0pocjEnk8PfFHPyZ1","level":3,"time":"2017-09-12T19:15:07+00:00","remoteAddr":"201.6.122.208","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/tmp\/ocr_lbBkA7.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}
{"reqId":"Sp50k1Z7qiSwlZI9uTlD","level":3,"time":"2017-09-12T19:31:01+00:00","remoteAddr":"201.6.122.208","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/tmp\/ocr_HGBi7Y.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}
{"reqId":"qcFLki3S9ZQoc2wOeNuO","level":3,"time":"2017-09-12T19:40:03+00:00","remoteAddr":"191.209.21.210","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/tmp\/ocr_uj7JOa.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}
{"reqId":"tpjHVYNAF4fi6yGWcVup","level":3,"time":"2017-09-12T19:42:07+00:00","remoteAddr":"191.209.21.210","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/tmp\/ocr_MDEOq4.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}

Other relevant information (inside ocr docker):
Mount -l:
/dev/sda on /home/node/data type ext4 (rw,relatime,discard,data=ordered)
/dev/vda1 on /home/node/output type ext4 (rw,relatime,data=ordered)

df -h:
Filesystem      Size  Used Avail Use% Mounted on
none             29G  5.0G   24G  18% /
tmpfs           496M     0  496M   0% /dev
tmpfs           496M     0  496M   0% /sys/fs/cgroup
/dev/vda1        29G  5.0G   24G  18% /etc/hosts
shm              64M     0   64M   0% /dev/shm
/dev/sda         49G  187M   47G   1% /home/node/data

('/dev/sda' in this list of docker is my HDD exclusive to data)

Relevant information in host server:
root@Core01:~# ls -la /var/www/html/tmp/
total 8
drwxr-xr-x  2 www-data www-data 4096 Sep 12 18:42 .
drwxr-xr-x 18 www-data www-data 4096 Sep 11 18:21 ..
-rw-------  1 www-data www-data    0 Sep 12 17:24 ocr_NGqTSA.txt

root@Core01:~# ls -la /mnt/volume-nyc1-01/files/
total 684
drwxrwx---  5 www-data www-data   4096 Sep 11 12:36 .
drwxr-xr-x  4 root     root       4096 Aug 29 18:17 ..
drwxr-xr-x  5 www-data www-data   4096 Aug 31 14:48 admin
drwxr-xr-x 10 www-data www-data   4096 Sep  4 12:50 appdata_ocelfraw3gi0
drwxr-xr-x  2 www-data www-data   4096 Aug 29 18:33 files_external
-rw-r--r--  1 www-data www-data    324 Sep 11 18:04 .htaccess
-rw-r--r--  1 www-data www-data      0 Sep 11 18:04 index.html
-rw-r-----  1 www-data www-data 670873 Sep 12 19:42 nextcloud.log
-rw-r--r--  1 www-data www-data      0 Sep 11 18:04 .ocdata

**Comment by caiofacchinato** [comment]
For testing purpose, i will try another test in new environment with history of commands:
1- `apt install apache2 libapache2-mod-php7.0 php7.0-gd php7.0-json php7.0-mysql php7.0-curl php7.0-mbstring php7.0-intl php7.0-mcrypt php-imagick php7.0-xml php7.0-zip mariadb-server`
2- `a2enmod rewrite && a2enmod headers && a2enmod env && a2enmod dir && a2enmod mime && systemctl restart apache2`
3- _(configure mariadb etc, download nc, unzip, chown to www-data, create tmp/ in /var/www/html & add this in tempdirectory in config.php, apt install php-redis and docker.io)_
4- on redis/: `docker build --build-arg PASSWORD="OCR" -t root/redis .`
5- on worker/:  _docker build --build-arg LANGUAGES="tesseract-ocr-por" -t root/ocr ._
6- `docker build --build-arg LANGUAGES="tesseract-ocr-por" -t root/ocr .`
7- `docker run --name ocr --network=isolated_ocr -e "REDIS_PASSWORD=OCR" -e "NODE_ENV=production" -e "REDIS_HOST=redis" -e "REDIS_PORT=6379" -e "REDIS_DB=0" -v /var/www/html/data:/home/node/data:ro -v /var/www/html/tmp:/home/node/output -d root/ocr`
8- _enable app, configure with defaults and try OCR one JPG._
9- root@ubuntu-Server-01:/var/www/html# docker logs 2985c64d1c1a
2017-09-13T16:35:59.839Z - error: Job errored: ENOENT: no such file or directory, unlink '/home/node/output//var/www/html/tmp/ocr_I65lZT.txt'
2017-09-13T16:36:31.448Z - error: Job errored: ENOENT: no such file or directory, unlink '/home/node/output//var/www/html/tmp/ocr_Iix3Af.txt'
2017-09-13T16:36:45.459Z - error: Job errored: ENOENT: no such file or directory, unlink '/home/node/output//var/www/html/tmp/ocr_LqU4fP.txt'
10- root@ubuntu-Server-01:/var/www/html# tail data/nextcloud.log
{"reqId":"BA1bMGmGrWoCPYOnQ7xf","level":3,"time":"2017-09-13T16:36:26+00:00","remoteAddr":"192.168.1.107","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/html\/tmp\/ocr_I65lZT.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}
{"reqId":"lQ9VGTvbomQoRhsF1y2Y","level":3,"time":"2017-09-13T16:36:36+00:00","remoteAddr":"192.168.1.107","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/html\/tmp\/ocr_Iix3Af.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}
{"reqId":"8eWJftjjt4giVCdwLdkY","level":3,"time":"2017-09-13T16:36:50+00:00","remoteAddr":"192.168.1.107","user":"admin","app":"no app in context","method":"GET","url":"\/index.php\/apps\/ocr\/status","message":"ENOENT: no such file or directory, unlink '\/home\/node\/output\/\/var\/www\/html\/tmp\/ocr_LqU4fP.txt'","userAgent":"Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/60.0.3112.113 Safari\/537.36","version":"12.0.0.29"}

**Comment by janis91** [comment]
It seems that there is something wrong with your symlinking inside of the docker container. The path `/home/node/output//var/www/html/tmp/ocr_Rk8oW3.txt` is obviously corrupted, because of two `/` following another. I think maybe your commands to create the docker container (docker run) is not the right one.. I just can guess, because for me and plenty others this works just fine.

**Comment by Braintelligence** [comment]
I have the same problem getting a double slash.
Strange enough if I docker exec -it into the container I see everything correct in /home/node/output. It definitely is the tempdirectory that I set.

Any leads?

**Comment by janis91** [comment]
Thanks for more input, now I can at least try it by myself and try to reproduce this.

**Comment by mathiasconradt** [comment]
I am having the same problems/error messages. I use the Docker approach, in fact I have everything running in Docker on Mac OS X (Nextcloud, OCR, Redis each in a Docker container).

When I try to OCR a document, I get messages like this in the OCR docker log:

`error: Job errored: ENOENT: no such file or directory, unlink '/home/node/output/ocr_8lHRca.txt'`

Will provide more details later on about my environment, currently still researching possible sources of this error.

**Comment by mathiasconradt** [comment]
@janis91 @Braintelligence @caiofacchinato I did some debugging and was able to fix the issue for me.

The problems seems to be in the `OcrProcessingService` class (`ocr-processing.service.ts`), in the `execTesseract` method, especially this line:

```this.unlinkSync(`${Configuration.OUTPUT_PATH}/${job.tempFile}`);```

When I debug the two variables, I get these values:

```configuration.OUTPUT_PATH: /home/node/output```
```job.tempFile: /var/www/html/tmp/ocr_FUMxTV.txt```

So, obviously, something is wrong here. The concatenation in the `unlinkSync` parameter combines two abolute paths, which cannot be right. Not sure what the supposed case is here.

Just for testing, I replaced it with

```this.unlinkSync(`${Configuration.OUTPUT_PATH}/${job.tempFile.replace('/var/www/html/tmp/','')}`);```

and the OCR processing works fine.
